File: Week5 (Tree-2 [BST, MST], Heap, Red-Black Tree)/Red-Black Tree/osj-rbt(main).c
/*
int main() {
    // Create a new red-black tree
    rbtree *tree = new_rbtree();

    // Insert some values into the tree
    rbtree_insert(tree, 10);
    rbtree_insert(tree, 20);
    rbtree_insert(tree, 30);
    rbtree_insert(tree, 40);
    rbtree_insert(tree, 50);
    rbtree_insert(tree, 60);
    
    // Find and print the minimum value in the tree
    node_t *min = rbtree_min(tree);
    if (min != tree->nil) {
        printf("Minimum value: %d\n", min->key);
    } else {
        printf("The tree is empty.\n");
    }

    // Find and print the maximum value in the tree
    node_t *max = rbtree_max(tree);
    if (max != tree->nil) {
        printf("Maximum value: %d\n", max->key);
    } else {
        printf("The tree is empty.\n");
    }

    // Find a specific value
    int key_to_find = 30;
    node_t *found = rbtree_find(tree, key_to_find);
    if (found != tree->nil) {
        printf("Found value: %d\n", found->key);
    } else {
        printf("Value %d not found in the tree.\n", key_to_find);
    }

    // Convert the tree to an array and print it
    size_t n = 6; // Number of elements inserted
    key_t *arr = (key_t *)malloc(n * sizeof(key_t));
    rbtree_to_array(tree, arr, n);

    printf("Tree elements in order: ");
    for (size_t i = 0; i < n; ++i) {
        printf("%d ", arr[i]);
    }
    printf("\n");

    // Free the allocated array
    free(arr);

    // Delete a node from the tree
    node_t *node_to_delete = rbtree_find(tree, 40);
    if (node_to_delete != tree->nil) {
        rbtree_erase(tree, node_to_delete);
    }

    // Delete the entire tree
    delete_rbtree(tree);

    return 0;
}
*/

File: Week5 (Tree-2 [BST, MST], Heap, Red-Black Tree)/Red-Black Tree/osj-rbt(main).py
'''
def main():
    rbt = RBT()
    
    # 노드 삽입 테스트
    elements = [20, 15, 25, 10, 5, 1, 30, 35, 40]
    for el in elements:
        rbt.insert(el)

    print("Red-Black Tree 구조 (삽입 후):")
    rbt.print_tree()

    # 노드 삭제 테스트
    node_to_delete = rbt.search(rbt.root, 15)
    if node_to_delete != rbt.Tnil:
        rbt.delete(node_to_delete)
        print("\nRed-Black Tree 구조 (노드 15 삭제 후):")
        rbt.print_tree()
    else:
        print("노드 15를 찾을 수 없음.")

if __name__ == "__main__":
    main()
'''